src/ft2font_wrapper.cpp

```diff
@@ -341,353 +168,187 @@ const char *PyFT2Font_init__doc__ =
     "\n"
     "    .. warning::\n"
     "        This API is private: do not use it directly\n"
-    "\n"
-    "Attributes\n"
-    "----------\n"
-    "num_faces : int\n"
-    "    Number of faces in file.\n"
-    "face_flags, style_flags : int\n"
-    "    Face and style flags; see the ft2font constants.\n"
-    "num_glyphs : int\n"
-    "    Number of glyphs in the face.\n"
-    "family_name, style_name : str\n"
-    "    Face family and style name.\n"
-    "num_fixed_sizes : int\n"
-    "    Number of bitmap in the face.\n"
-    "scalable : bool\n"
-    "    Whether face is scalable; attributes after this one are only\n"
-    "    defined for scalable faces.\n"
-    "bbox : tuple[int, int, int, int]\n"
-    "    Face global bounding box (xmin, ymin, xmax, ymax).\n"
-    "units_per_EM : int\n"
-    "    Number of font units covered by the EM.\n"
-    "ascender, descender : int\n"
-    "    Ascender and descender in 26.6 units.\n"
-    "height : int\n"
-    "    Height in 26.6 units; used to compute a default line spacing\n"
-    "    (baseline-to-baseline distance).\n"
-    "max_advance_width, max_advance_height : int\n"
-    "    Maximum horizontal and vertical cursor advance for all glyphs.\n"
-    "underline_position, underline_thickness : int\n"
-    "    Vertical position and thickness of the underline bar.\n"
-    "postscript_name : str\n"
-    "    PostScript name of the font.\n";
-
-static int PyFT2Font_init(PyFT2Font *self, PyObject *args, PyObject *kwds)
+;
+
+static PyFT2Font *
+PyFT2Font_init(py::object filename, long hinting_factor = 8,
+               std::optional<std::vector<PyFT2Font *>> fallback_list = std::nullopt,
+               int kerning_factor = 0)
 {
-    PyObject *filename = NULL, *open = NULL, *data = NULL, *fallback_list = NULL;
-    FT_Open_Args open_args;
-    long hinting_factor = 8;
-    int kerning_factor = 0;
-    const char *names[] = {
-        "filename", "hinting_factor", "_fallback_list", "_kerning_factor", NULL
-    };
-    std::vector<FT2Font *> fallback_fonts;
-    if (!PyArg_ParseTupleAndKeywords(
-             args, kwds, "O|l$Oi:FT2Font", (char **)names, &filename,
-             &hinting_factor, &fallback_list, &kerning_factor)) {
-        return -1;
-    }
     if (hinting_factor <= 0) {
-      PyErr_SetString(PyExc_ValueError,
-                      "hinting_factor must be greater than 0");
-      goto exit;
+        throw py::value_error("hinting_factor must be greater than 0");
     }
 
+    PyFT2Font *self = new PyFT2Font();
+    self->x = NULL;
```

Comment:
Ideally we'd use the C++ `nullptr` instead of `NULL`, but as there are a number of `NULL`s already in this file let's leave it consistent with that.